java-util/src/main/java/io/druid/java/util/common/granularity/PeriodSegmentGranularity.java

```diff
@@ -0,0 +1,166 @@
+            @JsonProperty("timeZone") DateTimeZone tz
+    )
+    {
+        this.periodGranularity = new PeriodGranularity(period, origin, tz);
+        this.granularityType = GranularityType.get(period);
+    }
+
+    @JsonProperty("period")
+    public Period getPeriod()
+    {
+        return periodGranularity.getPeriod();
+    }
+
+    @JsonProperty("timeZone")
+    public DateTimeZone getTimeZone()
+    {
+        return periodGranularity.getTimeZone();
+    }
+
+    @JsonProperty("origin")
+    public DateTime getOrigin()
+    {
+        return periodGranularity.getOrigin();
+    }
+
+    @Override
+    public DateTimeFormatter getFormatter(Formatter type) {
+        switch (type) {
+            case DEFAULT:
+                return DateTimeFormat.forPattern(granularityType.getDefaultFormat());
+            case HIVE:
+                return DateTimeFormat.forPattern(granularityType.getHiveFormat());
+            case LOWER_DEFAULT:
+                return DateTimeFormat.forPattern(granularityType.getLowerDefaultFormat());
+            default:
+                throw new IAE("There is no format for type %s", type);
+        }
+    }
+
+    @Override
+    public DateTime increment(DateTime time) {
+        return new DateTime(periodGranularity.increment(time.getMillis()));
+    }
+
+    @Override
+    public DateTime decrement(DateTime time) {
+        return new DateTime(periodGranularity.decrement(time.getMillis()));
+    }
+
+    @Override
+    public DateTime truncate(DateTime time) {
+        return new DateTime(periodGranularity.truncate(time.getMillis()));
+    }
+
+    @Override
+    public DateTime toDate(String filePath, Formatter formatter) {
+        Integer[] vals = getDateValues(filePath, formatter);
+
+        DateTime date = GranularityType.getDateTime(granularityType, vals);
+
+        if (date != null) {
+            return truncate(date);
+        }
+
+        return null;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        PeriodSegmentGranularity that = (PeriodSegmentGranularity) o;
+
+        if (granularityType != that.granularityType) {
+            return false;
+        }
+        return periodGranularity.equals(that.periodGranularity);
+
+    }
+
+    @Override
+    public int hashCode() {
+        int result = granularityType.hashCode();
+        result = 31 * result + periodGranularity.hashCode();
+        return result;
+    }
+
+    @Override
+    public String toString() {
+        return "{type=period, " +
+               "period=" + getPeriod() +
+               ", timeZone=" + getTimeZone() +
+               ", origin=" + getOrigin() +
+               '}';
+    }
+
+    @Override
+    public void serialize(JsonGenerator jsonGenerator, SerializerProvider serializerProvider) throws IOException, JsonProcessingException {
```

Comment:
"Not supporting rolling upgrade for 0.10.0" is not optional.  Supporting rolling updates is always required, even if we make backwards incompatible code changes.